### ISSUE TITLE
Slow startup on large servers when using high-rez textures and "clean transparent textures" mode

### ISSUE THREAD

**Issue by VanessaE** [minetest/minetest#11513]
##### Minetest version

Git commit 4a3728d828fa8896b49e80fdc68f5d7647bf45b7

##### OS / Hardware

Operating system: Debian 11
CPU: AMD Phenom II X6 1055T, 2.8 GHz
GPU model: AMD R9 280X, `amdgpu` driver
OpenGL version: 4.60

##### Summary

While testing another issue, I had to reset one of my configs, after which I started Minetest and enabled HDX-256.  It only took about one minute or so to join one of my Dreambuilder servers, which surprised me greatly...  I thought to myself "whoa, they fixed the sign-on time!"... then I realized that I'd forgotten to enable trilinear filtering, mipmap, and anisotropic filtering.

So I signed out, enabled those settings, quit the client, re-started, and signed back on.  This time it took ages to finish, which is what I originally expected.

This slow sign-on has actually been the case for me for a rather long time, and is definitely a regression.  In the past, signing onto such a server with my usual HDX/filters/etc. config selected used to take a little over one minute.  That's a long time, sure, but now it takes between 2 and 5 minutes (usually in the low end of that range).

All steps of signing on seem to be slower, but the bulk of the extra time is spent in the "Initializing nodes" step.  During that step, the client will work its way through a portion of the 0-100% progress range, then it sorta gets "stuck" for a long time on one percent value, after which it'll work through more of the progress, get "stuck" for a while again, etc., as though there are a handful of nodes (out of some 20'000) being evaluated that by themselves take many times longer to process than they should.

Since the slow sign-on only happens with the aforementioned filters enabled, I'm guessing there's been a performance regression in the process that's used scale the textures to generate the mipmaps.

I did not do a thorough test to track down the specific offending setting, but then again, one would normally always enable all three filters, or none of them.

Some of my other players also report that sign-ons take a few minutes for them as well for the Dreambuilder servers, though I can't be sure of their settings, texture pack, etc.

##### Steps to reproduce

* Start with a clean config
* download HDX-256 and enable it, but change no other settings
* sign onto one of my Dreambuilder servers (such as "Traintable", daconcepts.com port 30004), or maybe just use singleplayer mode with that game
* note the sign-on time, and in particular how long the "Initializing nodes" step takes
* sign off
* enable trilinear, mipmap, and anisotropic mode
* quit Minetest and restart it
* sign on again, note the sign-on time.  The "Initializing nodes" step should now take way longer than before.

You might give this config file a go: [minetest.conf-hires.zip](https://github.com/minetest/minetest/files/6935396/minetest.conf-hires.zip)

You'll probably have to disable and then re-enable HDX, since the path won't be the same as on my box.  Also note that after writing most of the above, I also changed a few other settings back to what they should be, but that had no effect on sign-on time.  This file reflects those settings.

**Comment by hecktest** [comment]
Could you try it with each of these settings in isolation? Sampler state isn't likely to affect load times, but the generation of mipmaps can.

This may unfortunately be an inherent issue with loading lots of high-res textures; the GPU driver has to generate mipmaps for all of them as they're loaded. This involves downscaling the texture by half repeatedly until it reaches a 1x1 size. Some GPU drivers might be smart about this, others less so. I'll have to dig around in irrlicht and confirm when and how mips are generated.

If we already are using the GPU driver to generate mips, this is not something that can be fixed; we could mitigate it by storing the pre-generated mips as part of the asset cache. I'm not sure how the cache interacts with texture packs, however.

**Comment by VanessaE** [comment]
> This may unfortunately be an inherent issue with loading lots of high-res textures;

This slowness wasn't an issue in the past, as I said.  The size of HDX has hovered around 2000 or so textures for a number of years now (in fact, the tendency has been to reduce the count by reworking mods to use them more intelligently).

> If we already are using the GPU driver to generate mips

This doesn't seem too likely, given Minetest does barely anything else on the GPU :stuck_out_tongue:  Besides, I would assume this would store the completed mipmaps in GPU RAM, and mine doesn't have nearly enough RAM for that.

> we could mitigate it by storing the pre-generated mips as part of the asset cache

That's a pretty good idea in my opinion.  Surely it's faster to generate them once and store them as files and then reload them from there in future runs, than to generate them on-demand.

**Comment by hecktest** [comment]
GPU mipmap generation is something you do once, but it may take a bit of time for a couple thousand textures since they are processed sequentially. A lot depends on the driver here, this might be a driver regression out of our reach.

> the tendency has been to reduce the count by reworking mods to use them more intelligently

Using texture modifiers does not actually reduce the amount of GPU textures generated, and thus the time spent generating mips is about the same.

If your GPU does not have enough VRAM to store all the textures with their mips, and it isn't crashing either, that means it's sharing space with regular memory and possibly making things even slower.
Please test which one of the three settings is actually causing this so that I can rule out stupid things like the driver attempting to use rip-mapping.

**Comment by VanessaE** [comment]
>  Using texture modifiers [...]

Well I wasn't thinking of that, but rather things like redesigning the UV mapping on mesh models so that they can use simple "standard" images like default stone or default wood, rather than using a custom texture that has both in some custom layout.

> A lot depends on the driver here, this might be a driver regression out of our reach.

If it is a driver regression, I can accept that, but that doesn't absolve the engine from doing what it can to reduce startup time.  But, see below... something isn't right.

> 2000 textures 256x256 each should fit on a 1GB card

Note that there are 2552 textures in HDX-256 at the moment, and my R9 280X GPU has 3 GB on it, and that not all 2552 textures are actually being used (as there are a fair number of textures in HDX for mods that others use, but which I do not).

From what I've seen, textures seem to be decoded into some kind of raw format when loaded, and then stored in regular system RAM.

The maths seems like it should be, at the absolute worst:

256 * 256 * 4 (bytes per pixel) * 2552 * 1.3333 (mipmap cost) = about 892 MB

That agrees with your assumption, but it disagrees with reality, so there's a "hidden factor" somewhere...

> Please test which one of the three settings is actually causing this [...]

Okay.  Taking your suggestion a but further, I've tested the various settings individually and in combination.

Before starting, I cleared `~/.cache/minetest`, then quit my browser, shut off swap and turned it back on.  Then I started Minetest in singleplayer mode, with HDX-256 enabled, went into my usual test world, and then quit, to repopulate Minetest's cache.  I figure this way, there's no chance of garbage/corrupt cache data, nothing will be using a ton of RAM or CPU except Minetest, and nothing will be trying to thrash about in swap.

Then I launched a local server (command line) to that same world, so that server-related performance wouldn't affect the results, so I could concentrate on client behavior.

Then for each test, I edited and wrote-out the config file (so that I didn't have to bother with Minetest's menus), then started Minetest and measured the server join time with a stopwatch, from the moment I clicked the "Connect" button to the moment the world appeared.  I made sure to keep an eye on the system's overall RAM usage as well, just to make sure it wasn't diving into swap.  I rounded up to the next integer second and the next 0.1 GB, and exited Minetest entirely after taking each reading.

Here are my results:

![image](https://user-images.githubusercontent.com/1733537/128689554-6e96bf85-6e6e-4706-9637-6855118f3fb7.png)

Obviously I tried to cover every valid combination of these settings - maybe this is overkill, but I guess it's better to have too much data than not enough.

For comparison, the client takes about 18 seconds and uses 1.1 GB of RAM to connect to that server process, when using my usual non-HDX config.  So, Minetest is using between 1.7 and 4.7 GB more RAM just for the bigger textures, depending on which combo of settings you compare it with.

I'm sure there's some margin of error here since I was doing this all "by hand", plus I had my PC's music player running, and I was making tweaks to this text (in a text editor, natch) while waiting, but I doubt that error is worse than ±2s or so.

It's pretty clear from the table that mipmap itself is not where the problem lies, as I can reliably reproduce the slow startup if I turn off everything except "clean transparent textures" (this combo is highlighted in the table above).

It should be noted that during all of those long startups, `top` showed the Minetest client process keeping one CPU core pegged at 100%, "Running" status. This suggests that there's some generic image scaling function in use that performs poorly under some circumstances.  Wouldn't high CPU load be shown as part of some other process if it were something in the GPU pipeline *outside* of Minetest (such as X itself or one of the kernel's helper processes)?

> VBOs might increase that requirement

As far as I know, VBOs only relate to world geometry, not texture behavior.  Way back when the pull request to add VBOs came about, the increase in RAM usage was minimal (in fact I don't think I even noticed, as I as focused on rendering performance at the time).

VBO has enabled in my config throughout all my tests.

> [...] so that I can rule out stupid things like the driver attempting to use rip-mapping.

What is "rip-mapping"?  I've never heard of this.

**Comment by sfan5** [comment]
> Obviously I tried to cover every valid combination of these settings - maybe this is overkill, but I guess it's better to have too much data than not enough.

FYI you could have saved yourself part of the tests, enabling mipmapping automatically enables "clean transparent textures" since a while.

**Comment by Emojigit** [comment]
That's unavoidable. Minetest have to load all files into the RAM first. The minimap uses some functions to get each node's color, which is (afaik) required to read each textures.

The second time you can access the server very fast is because (afaik) files are in the RAM, and (i guess) you have enough spaces to let Linux to cache those files even though minetest is closed.

**Comment by sfan5** [comment]
I recently changed the "clean transparent" algorithm and made it do more (fill the whole texture). For small textures that's okay, for large ones it's not necessary and also time consuming. The code already has an iteration limit for this reason:
https://github.com/minetest/minetest/blob/eefa39e47b4ef1f6a51854461efee99777ae20fd/src/client/imagefilters.cpp#L98-L100
Now 50 is probably way too high. How do the loadtimes look when you change this to 10?
You can also comment out [the call in tile.cpp](https://github.com/minetest/minetest/blob/master/src/client/tile.cpp#L1607) to compare times without the filter forced on.

Other possibility is that the algorithm is just slow for whatever other reason. Hard to say without performance measurements (ideally a flamegraph).

**Comment by hecktest** [comment]
Thanks for the exhaustive test.
I actually forgot that we tie anisotropic filtering to mipmaps. By itself, AF does not have the power to cause something like this because it's a sampler setting akin to filtering.
From the table it appears that texture filtering settings do not have any bearing on the startup time, and every time the startup was short, mipmaps were off. Other settings are unlikely to have any influence.

**Comment by VanessaE** [comment]
@sfan5
> Now 50 is probably way too high. How do the loadtimes look when you change this to 10?

Using the same world and client/server separation as before, with my usual HDX config (and hence, "ENABLE ALL THE FILTERS!" :grin:), setting that cap at 10 reduced the join time from the previous peak of 2:24 down to 1:39.

Just for grins, I dropped the cap to 2.  That reduced the join time to 1:14, a full minute less than the peak.

For both tests, I checked a bunch of different nodes in the world that use transparency (leaves and other foliage, glass, doors with windows in them, etc.), both high- and low-resolution. Note that low-resolution textures are still automatically scaled-up per the "minimum texture size" feature (which is set to 256 to match HDX).

I saw nothing wrong with the reduced caps.  Even at 2, everything looked fine to me.

I also tried commenting-out that function call like you suggested (and disabled it in my config).  That reduced the join time to 1:07.

However, with it disabled entirely like that, some textures definitely do have some darkness around their opaque-transparent edges (for example, `vines:side_middle` from plantlife modpack vines mod, which uses `vines_side_middle.png`).

I also did a quick check with the cap set to 2, using my usual low-resolution config, but enabling trilinear, mipmap+aniso, 4x antialising, and clean transparent (so, the same as I use in HDX mode), just to what effect that cap has.  I saw nothing wrong anywhere.

Since it seems to work fine with both low-resolution and high-resolution textures, I have to wonder if it wouldn't be sufficient to just reduce the cap to 2 and call it a day?  Of course, I would definitely suggest others take a close look at how that affects textures on their machines.  More eyes and all that.

Also, an important note:

This is a bit of a tangent, but in the times listed above and in my chart from before, a good solid 15 seconds in each test is spent in the "Media" phase.

Since everything's on my local PC (so no Internet delays), at first I was just going to time spent loading media, but then I noticed that there's very little disk activity during this period, and the machine runs from a decent SSD anyway, so I wonder if there's some performance gains to be had in the "Media" phase.
